### PR TITLE
Fixes #32837 - drop unwanted URI parsing from extension

### PR DESCRIPTION
**Pull request by lzap**
Ruby Net::HTTP object does not contain URI, it contains address and port attributes which is documented to only old IP address or hostname and port, not URI or URL at all.

In our NetHttpExtension we parse the address into URI and then extract hostname part from there. This works fine for IPv4, however it fails for IPv6 as addresses must be enclosed within square brackets. The patch is removing the not-needed conversion to URI and back.

See the issue for full backtrace.

### CODE REVIEW

File: lib/foreman/http_proxy/net_http_extension.rb [hunk 1]
```diff
@@ -2,9 +2,8 @@ module Foreman
   module HttpProxy
     module NetHttpExtension
       def request(req, body = nil, &block)
-        host = URI.parse(@address).host
```
Comment (lzap) [inline lib/foreman/http_proxy/net_http_extension.rb:5]: Looking into Ruby code, it looks like host == `@address`.

https://github.com/ruby/ruby/blob/master/lib/net/http.rb#L671-L676

File: lib/foreman/http_proxy/net_http_extension.rb [hunk 2]
```diff
@@ -2,9 +2,8 @@ module Foreman
...
-        if proxy_http_request?(@proxy_address, host, @socket)
-          log_proxied_request("NetHttp", http_proxy, host)
+        if proxy_http_request?(@proxy_address, @address, @socket)
```
Comment (ekohl) [inline lib/foreman/http_proxy/net_http_extension.rb]: How did this ever work? The signature is here:
https://github.com/theforeman/foreman/blob/44764290e760111db7c41b859cc242a74e3516ed/lib/foreman/http_proxy.rb#L16

So the third argument is a schema, but a socket is passed in? How is that ever true?

**Review by ekohl — COMMENTED** [review]
Looks like you don't even need to use the internal variable:
https://github.com/ruby/ruby/blob/9d96837dbd1887d04f5ff7c2a1f0a27d7339133a/lib/net/http.rb#L727-L728
But given that it was there before, this indeed looks redundant and even potentially breaking as you've found out.

That said, I really wonder about `@socket` here. But maybe it's just Friday and hot so I can't think clearly anymore.

File: lib/foreman/http_proxy/net_http_extension.rb (hunk 1 above)
Comment (tbrisker) [reply lib/foreman/http_proxy/net_http_extension.rb:5]: Might this be an attempt to verify the address is a valid URI?

Comment (lzap) [reply lib/foreman/http_proxy/net_http_extension.rb:5]: Oh no, I failed to explain this miserably. Let me try again: at no point, `@address` contains a URI/URL. It contains either hostname or IP address. And the reason we see the error is that Ruby is a bit inconsistent here - it does parse invalid URI of `1.2.3.4` but it does not parse the same with IPv6 address:

```
>> URI.parse("1.2.4.5")
=> #<URI::Generic 1.2.4.5>
>> URI.parse("2a00:4bc0:1010:1102:0:b19:7175:babe")
Traceback (most recent call last):
        7: from /home/lzap/.rbenv/versions/2.7.2/bin/irb:23:in `<main>'
        6: from /home/lzap/.rbenv/versions/2.7.2/bin/irb:23:in `load'
        5: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        4: from (irb):2
        3: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/uri/common.rb:234:in `parse'
        2: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/uri/rfc3986_parser.rb:73:in `parse'
        1: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/uri/rfc3986_parser.rb:67:in `split'
URI::InvalidURIError (bad URI(is not URI?): "2a00:4bc0:1010:1102:0:b19:7175:babe")
>> URI.parse("[2a00:4bc0:1010:1102:0:b19:7175:babe]")
Traceback (most recent call last):
        8: from /home/lzap/.rbenv/versions/2.7.2/bin/irb:23:in `<main>'
        7: from /home/lzap/.rbenv/versions/2.7.2/bin/irb:23:in `load'
        6: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        5: from (irb):2
        4: from (irb):3:in `rescue in irb_binding'
        3: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/uri/common.rb:234:in `parse'
        2: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/uri/rfc3986_parser.rb:73:in `parse'
        1: from /home/lzap/.rbenv/versions/2.7.2/lib/ruby/2.7.0/uri/rfc3986_parser.rb:67:in `split'
URI::InvalidURIError (bad URI(is not URI?): "[2a00:4bc0:1010:1102:0:b19:7175:babe]")
```

File: lib/foreman/http_proxy/net_http_extension.rb (hunk 2 above)
Comment (lzap) [reply lib/foreman/http_proxy/net_http_extension.rb]: Hah lemme see, it's Monday so I should be fresh now.

**Comment by lzap** [comment]
> you don't even need to use the internal variable

Agreed, will you carefully test the change in that case? I was aiming towards minimum possible change to be honest, this code does look fragile. Confirm and I will do the refactoring (should not hurt but).

File: lib/foreman/http_proxy/net_http_extension.rb (hunk 2 above)
Comment (lzap) [reply lib/foreman/http_proxy/net_http_extension.rb]: Correction, it's Tuesday. Yay! :-)

Anyway, it looks like @socket is either:

1. `nil` if you are doing non-persistent connection, which by default are all connections unless you call Net::HTTP.start with a block
2. `false` in all other cases:

         ['http', 'https'].include?(Net::BufferedIO.new(StringIO.new))

Conclusion: this has never worked, at least the condition there.

Comment (lzap) [reply lib/foreman/http_proxy/net_http_extension.rb]: To debug this:

```
irb

require "pry"
module NetHttpExtension
   def request(req, body = nil, &block)
      binding.pry
      super(req, body, &block)
    end
 end
Net::HTTP.class_eval do
  prepend NetHttpExtension
end

[9] pry(#<Net::HTTP>)> @socket
=> #<Net::BufferedIO io=#<TCPSocket:0x000000000112ec28>>
[10] pry(#<Net::HTTP>)> ['http', 'https'].include?(@socket)
=> false
```

The correct solution is: `req.uri.scheme.start_with?("http")`

Comment (ekohl) [reply lib/foreman/http_proxy/net_http_extension.rb]: Just wondering: when is it ever not a HTTP(S) scheme? When it's some unix socket? I found https://github.com/puppetlabs/net_http_unix so I guess technically that could be it.

Comment (lzap) [reply lib/foreman/http_proxy/net_http_extension.rb]: I have no idea to be honest, this was added from the day one across all three implementations (Net::HTTP, RestClient, Excon) in 7f8fd14d7a7fb4030bc2e6c36a297b4aa864b7ba maybe @bastilian would put some light on it?

Sebastian a question for you: why the code passed URI scheme to the `proxy_http_request?` method? Aren't all three libraries purely for HTTP and therefore this check is not needed? Two are passing scheme correctly, but Net::HTTP is passing `@socket` instead which is an IO object and not URI scheme.

**Comment by lzap** [comment]
It is green now.

**Comment by ezr-ondrej** [comment]
We are now not validating the URI (previously we did by parsing it). Is this an issue in your opinion @lzap ?

File: lib/foreman/http_proxy/net_http_extension.rb [hunk 3]
```diff
@@ -2,9 +2,8 @@ module Foreman
   module HttpProxy
     module NetHttpExtension
       def request(req, body = nil, &block)
-        host = URI.parse(@address).host
-        if proxy_http_request?(@proxy_address, host, @socket)
-          log_proxied_request("NetHttp", http_proxy, host)
+        if proxy_http_request?(@proxy_address, address, req&.uri&.scheme)
```
Comment (ekohl) [inline lib/foreman/http_proxy/net_http_extension.rb]: I still don't understand the need for `proxy_http_request?` because Net::HTTP already has a method `proxy?` which really does the determination if a proxy is used. Having different logic just makes things confusing.

File: lib/foreman/http_proxy/net_http_extension.rb [hunk 4]
```diff
@@ -2,9 +2,8 @@ module Foreman
...
+          log_proxied_request("NetHttp", http_proxy, address)
           @proxy_address = URI.parse(http_proxy)
```
Comment (ekohl) [inline lib/foreman/http_proxy/net_http_extension.rb:7]: This is also really confusing to me. If a connection was already started (can be checked with `started?`) this doesn't have any effect.

**Review by ekohl — COMMENTED** [review]
When `request` is called, `Net::Http` is already configured to connect to a host. If a host is invalid, validation should happen prior to it. An example:

```ruby
Net::HTTP.start(host, port, proxy_addr, proxy_port) do |http|
  # Here it will already be connected
  http.request(req) # This would have validated the host
end
```

So the validation would always have been too late in that case.

To me it's unclear how this would ever reliably work. Given that previously it was dead code (the condition could never be true), we will be activating new code here. So my concern isn't with just validation, but the whole module. To me ripping out the monkey patching is on the table.

**Comment by lzap** [comment]
I believe this is ready for review.

> We are now not validating the URI (previously we did by parsing it). Is this an issue in your opinion @lzap ?

It was not validating anything, that is the point. The `@address` only contained hostname/ip so it was always valid.

I agree with @ekohl this monkey patching is terrible, however, it is not my intention to refactor this at this point. I just want to fix the bug.

**Comment by ares** [comment]
I read through the PR and I'm far from understanding the need or safetyness of the change. I understood that we're dropping `URI.parse` as it can fail on IPv6. So that's being dropped and we are changing what we pass to `proxy_http_request?` we used before.

Second problem is, the use of `proxy_http_request?` was always wrong, because the third argument is supposed to be scheme. That is IMHO unrelated and there are two options.
1) we fix it here and it requires better explanation of what happens in that case
2) we don't touch it here and just modify host -> address

I think that's contributor choice, however we'd know we're leaving broken stuff behind.

If we want to go with option 2), @ekohl could you please summarize what is the concern with the actual version, meaning passing the correct scheme? I didn't get it from your example when you try to demonstrate the validation being too late. Is that related to the actual change here or the NetHttpExtension module in general? The other two comments seems unrelated or does not point to a specific issue with the changed code. It talks about existing code being confusing. I think we all agree to that.

**Comment by lzap** [comment]
I agree with Marek, I removed the unrelated scheme change, the patch is as small as possible to fix the bug.

I created a separate ticket for the scheme issue and possibly whole refactoring: https://projects.theforeman.org/issues/33247

File: lib/foreman/http_proxy/net_http_extension.rb [hunk 5]
```diff
@@ -2,9 +2,8 @@ module Foreman
   module HttpProxy
     module NetHttpExtension
       def request(req, body = nil, &block)
-        host = URI.parse(@address).host
-        if proxy_http_request?(@proxy_address, host, @socket)
-          log_proxied_request("NetHttp", http_proxy, host)
+        if proxy_http_request?(@proxy_address, address, @socket)
+          log_proxied_request("NetHttp", http_proxy, address)
```
Comment (tbrisker) [inline lib/foreman/http_proxy/net_http_extension.rb:6]: if i'm not mistaken, address is an instance variable:
```suggestion
        if proxy_http_request?(@proxy_address, @address, @socket)
          log_proxied_request("NetHttp", http_proxy, @address)
```

Comment (lzap) [reply lib/foreman/http_proxy/net_http_extension.rb:6]: You are not, but there is also a reader in the Net::HTTP class:

```ruby
      # The DNS host name or IP address to connect to.
      attr_reader :address
```

I thought I would stick with it. No writer tho.

**Review by ekohl — APPROVED** [review]
I still think we need to evaluate this entire class because I still have a theory that if we drop the entire extension, nothing would be lost. However, I'm ok merging this since it at least remove one breakage it does cause.

**Comment by upadhyeammit** [comment]
Thank you for reviews! Merging this now and will cherry pick into 3.0.

**Comment by upadhyeammit** [comment]
cp done into 3.0 https://github.com/theforeman/foreman/commit/990a587ee11cd16aac6c48517225c8877f1dbdca
